lib/internal/fs/glob.js

```diff
@@ -169,212 +208,330 @@ class Glob {
+    if (seen) {
       return;
     }
 
+    if (pattern.isFirst()) {
+      const shouldFinish = this.addPatternsOnFirst(pattern);
+
+      if (shouldFinish) {
+        return;
+      }
+    }
+
+    const fullpath = resolve(this.#root, path);
+    const stat = await this.#cache.stat(fullpath);
+    const last = pattern.last;
+    const isDirectory = stat?.isDirectory() || (stat?.isSymbolicLink() && pattern.hasSeenSymlinks);
+    const isLast = pattern.isLast(isDirectory);
+
+
     if (isLast && typeof pattern.at(-1) === 'string') {
       // Add result if it exists
       const p = pattern.at(-1);
-      const stat = this.#cache.statSync(join(fullpath, p));
+      const stat = await this.#cache.stat(join(fullpath, p));
       if (stat && (p || isDirectory)) {
-        ArrayPrototypePush(this.#results, join(path, p));
+        yield join(path, p);
       }
       if (pattern.indexes.size === 1 && pattern.indexes.has(last)) {
         return;
       }
     } else if (isLast && pattern.at(-1) === lazyMinimatch().GLOBSTAR &&
-     (path !== '.' || pattern.at(0) === '.' || (last === 0 && stat))) {
+      (path !== '.' || pattern.at(0) === '.' || (last === 0 && stat))) {
       // If pattern ends with **, add to results
       // if path is ".", add it only if pattern starts with "." or pattern is exactly "**"
-      ArrayPrototypePush(this.#results, path);
+      yield path;
     }
 
     if (!isDirectory) {
       return;
     }
 
-    let children;
     const firstPattern = pattern.indexes.size === 1 && pattern.at(pattern.indexes.values().next().value);
     if (typeof firstPattern === 'string') {
-      const stat = this.#cache.statSync(join(fullpath, firstPattern));
+      const stat = await this.#cache.stat(join(fullpath, firstPattern));
       if (stat) {
         stat.name = firstPattern;
-        children = [stat];
-      } else {
-        children = [];
+        yield* this.addFromChildren(stat, path, fullpath, pattern, last, isLast);
       }
-    } else {
-      children = this.#cache.readdirSync(fullpath);
+
+      return;
     }
 
-    for (let i = 0; i < children.length; i++) {
-      const entry = children[i];
-      const entryPath = join(path, entry.name);
-      this.#cache.addToStatCache(join(fullpath, entry.name), entry);
+    const children = await this.#cache.readdir(fullpath);
+
+    for (const entry of children) {
+      yield* this.addFromChildren(entry, path, fullpath, pattern, last, isLast);
+    }
+  }
+
+  * addFromChildren(entry, path, fullpath, pattern, last, isLast) {
+    const entryPath = join(path, entry.name);
+    this.#cache.addToStatCache(join(fullpath, entry.name), entry);
+
+    const subPatterns = new SafeSet();
+    const nSymlinks = new SafeSet();
+    for (const index of pattern.indexes) {
+      // For each child, check potential patterns
+      if (this.#cache.seen(entryPath, pattern, index) || this.#cache.seen(entryPath, pattern, index + 1)) {
+        // TODO - fix this not finish the addSubPatterns
```

Comment:
@MoLow any JS trick to finish the called generator?

Reply:
test the return value of `addFromChildren`
```js
for (const entry of children) {
	const result = this.addFromChildren(entry, path, fullpath, pattern, last, isLast);
	if (result === null) {
		return;
	}
	yield result;
}
```

Reply:
`addFromChildren` is a generator by itself so I need to get the value from it but thought if have another way